In [9]:
import pandas as pd

df = pd.read_excel("Dataset for Data Analytics.xlsx")

df.head()

,OrderID,Date,CustomerID,Product,Quantity,UnitPrice,ShippingAddress,PaymentMethod,OrderStatus,TrackingNumber,ItemsInCart,CouponCode,ReferralSource,TotalPrice
0,ORD200000,2023-01-04,C72649,Monitor,5,570.62,928 Main St,Debit Card,Shipped,TRK37947903,7,SAVE10,Instagram,2853.10
1,ORD200001,2024-08-23,C75739,Phone,2,151.35,823 Main St,Online,Shipped,TRK91186779,3,SAVE10,Referral,302.70
2,ORD200002,2024-02-27,C81728,Tablet,5,550.68,512 Main St,Credit Card,Cancelled,TRK42903982,8,FREESHIP,Email,2753.40
3,ORD200003,2023-10-15,C33540,Chair,1,273.19,275 Main St,Debit Card,Returned,TRK62788070,5,SAVE10,Facebook,273.19
4,ORD200004,2025-05-08,C81840,Printer,4,626.01,668 Main St,Online,Delivered,TRK29241424,8,SAVE10,Email,2504.04


In [10]:
%pip install openpyxl

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 25.3 -> 26.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [11]:
df.shape

(1200, 14)

In [12]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1200 entries, 0 to 1199
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   OrderID          1200 non-null   str           
 1   Date             1200 non-null   datetime64[us]
 2   CustomerID       1200 non-null   str           
 3   Product          1200 non-null   str           
 4   Quantity         1200 non-null   int64         
 5   UnitPrice        1200 non-null   float64       
 6   ShippingAddress  1200 non-null   str           
 7   PaymentMethod    1200 non-null   str           
 8   OrderStatus      1200 non-null   str           
 9   TrackingNumber   1200 non-null   str           
 10  ItemsInCart      1200 non-null   int64         
 11  CouponCode       891 non-null    str           
 12  ReferralSource   1200 non-null   str           
 13  TotalPrice       1200 non-null   float64       
dtypes: datetime64[us](1), float64(2), int64(2), str(9)


In [13]:
df.isnull().sum()

OrderID              0
Date                 0
CustomerID           0
Product              0
Quantity             0
UnitPrice            0
ShippingAddress      0
PaymentMethod        0
OrderStatus          0
TrackingNumber       0
ItemsInCart          0
CouponCode         309
ReferralSource       0
TotalPrice           0
dtype: int64

In [14]:
df.duplicated().sum()

np.int64(0)

In [15]:
df["CouponCode"].value_counts(dropna=False)

CouponCode
FREESHIP    313
NaN         309
WINTER15    292
SAVE10      286
Name: count, dtype: int64

In [16]:
df["CouponCode"] = df["CouponCode"].fillna("No Coupon")

In [17]:
df["CouponCode"].value_counts()

CouponCode
FREESHIP     313
No Coupon    309
WINTER15     292
SAVE10       286
Name: count, dtype: int64

## 1. Handling Missing Values

The dataset was checked for missing values using `isnull().sum()`.

The `CouponCode` column contained 309 missing values. These missing values were investigated using `value_counts()` to understand whether they represented an actual data issue.

Since the `CouponCode` field is optional, the missing values were interpreted as orders where no coupon was used. Therefore, the missing values were replaced with `"No Coupon"` rather than removing the affected records.

After cleaning, the `CouponCode` column contained four categories: `FREESHIP`, `WINTER15`, `SAVE10`, and `No Coupon`.

In [18]:
df["OrderID"].nunique()

1200

## 2. Checking for Duplicate Orders

The `OrderID` column was checked to determine whether each order had a unique identifier.

The dataset contains 1,200 records and 1,200 unique OrderIDs. This confirms that there are no duplicate OrderIDs in the dataset.

In [19]:
df["PaymentMethod"].value_counts()

PaymentMethod
Online         258
Cash           246
Credit Card    234
Debit Card     232
Gift Card      230
Name: count, dtype: int64

In [20]:
df["OrderStatus"].value_counts()

OrderStatus
Cancelled    250
Returned     247
Pending      237
Shipped      235
Delivered    231
Name: count, dtype: int64

## 3. Checking Categorical Data Consistency

The `PaymentMethod` and `OrderStatus` columns were inspected for inconsistent spelling, capitalization, and unexpected category values.

The `PaymentMethod` column contained five consistent categories: `Online`, `Cash`, `Credit Card`, `Debit Card`, and `Gift Card`.

The `OrderStatus` column contained five consistent categories: `Cancelled`, `Returned`, `Pending`, `Shipped`, and `Delivered`.

No inconsistent or invalid values were identified in either column, so no cleaning was required.

In [21]:
df["ReferralSource"].value_counts()

ReferralSource
Instagram    259
Email        250
Google       241
Facebook     228
Referral     222
Name: count, dtype: int64

In [22]:
df["Quantity"].describe()

count    1200.000000
mean        2.945833
std         1.407557
min         1.000000
25%         2.000000
50%         3.000000
75%         4.000000
max         5.000000
Name: Quantity, dtype: float64

In [23]:
df["UnitPrice"].describe()

count    1200.000000
mean      356.412750
std       197.177146
min        11.390000
25%       186.062500
50%       364.210000
75%       521.570000
max       699.930000
Name: UnitPrice, dtype: float64

In [24]:
df["CalculatedTotal"] = df["Quantity"] * df["UnitPrice"]

In [25]:
df[["Quantity", "UnitPrice", "TotalPrice", "CalculatedTotal"]].head(10)

,Quantity,UnitPrice,TotalPrice,CalculatedTotal
0,5,570.62,2853.10,2853.10
1,2,151.35,302.70,302.70
2,5,550.68,2753.40,2753.40
3,1,273.19,273.19,273.19
4,4,626.01,2504.04,2504.04
5,2,245.86,491.72,491.72
6,1,664.42,664.42,664.42
7,5,149.55,747.75,747.75
8,2,134.28,268.56,268.56
9,4,509.38,2037.52,2037.52


In [26]:
(df["TotalPrice"] - df["CalculatedTotal"]).abs().max()

np.float64(4.547473508864641e-13)

## 4. Cleaning and Validation of Numerical Columns

The numerical columns `Quantity`, `UnitPrice`, and `TotalPrice` were inspected for missing, zero, negative, and inconsistent values.

The `Quantity` column contains values ranging from 1 to 5, with no zero or negative quantities identified. The `UnitPrice` column contains values ranging from 11.39 to 699.93, with no zero, negative, or missing prices found.

The `TotalPrice` column was validated by comparing it with the calculated value of `Quantity × UnitPrice`. The largest difference across all 1,200 records was approximately 4.55 × 10⁻¹³, which is effectively zero and is attributed to floating-point precision.

Therefore, no corrections were required for the numerical columns.

In [27]:
df["Date"].describe()

count                   1200
mean     2024-03-22 16:58:48
min      2023-01-01 00:00:00
25%      2023-08-03 18:00:00
50%      2024-03-23 00:00:00
75%      2024-11-08 12:00:00
max      2025-06-30 00:00:00
Name: Date, dtype: object

In [28]:
df["Date"].dt.year.value_counts().sort_index()

Date
2023    510
2024    459
2025    231
Name: count, dtype: int64

## 5. Cleaning and Validation of Date Data

The `Date` column was inspected for missing values, incorrect data types, and unusual date ranges.

All 1,200 records contain valid dates, and the column is correctly stored as a datetime data type. The dates range from January 1, 2023 to June 30, 2025.

The distribution of records across the years was also examined, with 510 orders in 2023, 459 orders in 2024, and 231 orders in 2025. The lower number of records in 2025 is reasonable because the dataset ends on June 30, 2025.

No date-related issues were identified, so no cleaning was required.

In [29]:
df["Product"].value_counts()

Product
Printer    181
Tablet     179
Chair      178
Laptop     173
Desk       170
Monitor    163
Phone      156
Name: count, dtype: int64

In [30]:
df["CustomerID"].head(10)

0    C72649
1    C75739
2    C81728
3    C33540
4    C81840
5    C37249
6    C83492
7    C41460
8    C26817
9    C31946
Name: CustomerID, dtype: str

In [31]:
df["OrderID"].head(10)

0    ORD200000
1    ORD200001
2    ORD200002
3    ORD200003
4    ORD200004
5    ORD200005
6    ORD200006
7    ORD200007
8    ORD200008
9    ORD200009
Name: OrderID, dtype: str

In [32]:
df["TrackingNumber"].head(10)

0    TRK37947903
1    TRK91186779
2    TRK42903982
3    TRK62788070
4    TRK29241424
5    TRK72976927
6    TRK96417362
7    TRK78809193
8    TRK61042692
9    TRK33478363
Name: TrackingNumber, dtype: str

In [33]:
df["ShippingAddress"].head(10)

0    928 Main St
1    823 Main St
2    512 Main St
3    275 Main St
4    668 Main St
5    934 Main St
6    986 Main St
7    706 Main St
8    904 Main St
9    102 Main St
Name: ShippingAddress, dtype: str

## 6. Cleaning and Validation of Text and Identifier Columns

The text and identifier columns were inspected for missing values, duplicate identifiers, inconsistent formatting, spelling differences, and unexpected values.

The `OrderID` column contains 1,200 unique identifiers following a consistent format beginning with `ORD`. The `CustomerID` and `TrackingNumber` values were also inspected and appeared to follow consistent formats.

The `Product` column contained seven consistently formatted product categories: `Printer`, `Tablet`, `Chair`, `Laptop`, `Desk`, `Monitor`, and `Phone`.

The `PaymentMethod`, `OrderStatus`, and `ReferralSource` columns contained consistent and valid categorical values with no obvious spelling or capitalization inconsistencies.

The `ShippingAddress` column contained no missing values, and the sampled addresses followed a consistent format.

No cleaning was required for these columns.

In [43]:
df.isnull().sum()

OrderID            0
Date               0
CustomerID         0
Product            0
Quantity           0
UnitPrice          0
ShippingAddress    0
PaymentMethod      0
OrderStatus        0
TrackingNumber     0
ItemsInCart        0
CouponCode         0
ReferralSource     0
TotalPrice         0
dtype: int64

## 7. Final Data Validation

After completing the cleaning process, the dataset was checked again for missing values to ensure that no missing data remained.

All 14 columns returned zero missing values. This confirms that the missing `CouponCode` values were successfully handled and that no additional missing values are present in the cleaned dataset.

In [44]:
df.shape

(1200, 14)

In [45]:
df.to_excel("Cleaned_Dataset_for_Data_Analytics.xlsx", index=False)